File: python/tuning.py
import numpy as np



def get_freqs(ref_freq, semitone, low_exp, high_exp):
    # freq = fundamental * SEMITONE^exponent
    return np.array([ref_freq * semitone ** n for n in range(low_exp, high_exp + 1)])

def calculate_linear_exnotes(self):
    count = self.frequencies.shape[0]

    interval = self.frequencies.copy()
    interval = (np.roll(self.frequencies,-1) - interval)/3
    o1 = self.frequencies.copy()
    o2 = o1 + interval
    o3 = o2 + interval

    output = np.array([o1, o2, o3]).transpose().reshape(count*3,1).flatten()
    return output[:output.shape[0]-2]

def calculate_nonlinear_exnotes(self):
    count = self.frequencies.shape[0]
    low = self.C0_exponent
    high = self.B8_exponent

    o1 = self.frequencies.copy()
    o2 = np.array([self.tuning_freq * self.semitone ** (n + 1/3) for n in range(low, high + 1)])
    o3 = np.array([self.tuning_freq * self.semitone ** (n + 2/3) for n in range(low, high + 1)])

    output = np.array([o1, o2, o3]).transpose().reshape(count * 3, 1).flatten()
    return output[:output.shape[0] - 2]

def get_index(freq, refFreq, semitone, minExponent):
    # the lowest note of consideration (i.e. C0_exponent) will be considered 0
    # this can then be used to lookup values from the note table by note name of frequency

    # reverse calculate_notes()
    semitone_to_exp = freq / refFreq

    exponent = np.log(semitone_to_exp) / np.log(semitone)

    # can't go below the lowest exponent / C0
    exponent = max(minExponent, exponent)
    offset = int(0.5 + exponent - minExponent)  # round up to the nearest exponent
    return offset

# integer based pitch class
def get_class_index(freq):

    offset = get_index(freq)
    index = offset % 12
    return index

def get_octave(freq):
    offset = get_index(freq)
    octave = offset//12
    return octave

def oct_ind_to_freq(freqs, octave, index):
    i = octave*12 + index
    return freqs[i]

def crush(exarr, final_size, shift = 1, factor = 3):
    output = np.zeros(final_size)

    for exind in range(exarr):
        ind = (exind + shift)//factor
        output[ind] = np.max(output[ind], exarr[exind])

File: Source/Tuning.cpp
#include "Tuning.h"

// Tuning calculations ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
float getEqualTemperamentFrequency(float refFreq, float semitone, int exponent)
{
    return refFreq * pow(semitone, exponent);
}

int getEqualTemperamentExponent(float freq, float refFreq, float semitone)
{
    float semitoneToExp = freq / refFreq;
    float exponent = std::log(semitoneToExp) / std::log(semitone); 
    return int(0.5 + exponent);
}

int getIndex(float freq, float refFreq, float semitone, int minExponent)
{
    int exponent = getEqualTemperamentExponent(freq, refFreq, semitone);
    exponent = std::max(minExponent, exponent);
    
    //make minExponent = 0 index in freq array
    exponent = exponent - minExponent; 
    return exponent;
}

//pitch class. I.e. C, D, G, G# etc. indices [0,11] for 12Tone Equal Temperament
int getClassIndex(float freq, float refFreq, float semitone, int minExponent, const strvec& noteNames)
{
    int index =  getIndex(freq, refFreq, semitone, minExponent );
    index = index % noteNames.size();
    return index;
}

std::string getClassName(float freq, float refFreq, float semitone, int minExponent, const strvec& noteNames)
{
    return noteNames[getClassIndex(freq, refFreq, semitone, minExponent, noteNames)];
}

int getOctave(float freq, float refFreq, float semitone, int minExponent)
{
    int index =  getIndex(freq, refFreq, semitone, minExponent);
    int octave = int(index/12);
    return octave;
}

float octIndToFreq(const fvec& frequencies, int octave, int pitch, int noteCount)
{
    int freqInd = octave*noteCount + pitch;
    return frequencies[freqInd];
}

fvec getOctFrequencies(const fvec& refFreqs, int octSize)
{
    fvec output(octSize, 0);

    for(int i = 0; i < output.size(); i++)
    {
        output[i] = refFreqs[i * octSize];
    }
    return output;
}

// fill frequency arrays ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
void setETFrequencies(fvec& freqs, float refFreq, float semitone, int lowExp, int highExp)
{
    int i = 0;
    for(int exp = lowExp; exp < highExp + 1; exp++)
    {
        freqs[i] = getEqualTemperamentFrequency(refFreq, semitone, exp);
        i++;
    }

}

void expandFrequencies(fvec& freqs, fvec& exFreqs)
{
    //linear interpolation between frequencies

    for(int i = 0; i<freqs.size() -1; i++)
    {
        int ind = 3*i;
        exFreqs[ind] = freqs[i];

        float interval = (freqs[i+1] - freqs[i] )/3.0;
        exFreqs[ind + 1] = exFreqs[ind] + interval;
        exFreqs[ind + 2] = exFreqs[ind] + 2*interval;
    }

    
    exFreqs[exFreqs.size()-1] = freqs[freqs.size() -1];

}

fvec crush(const fvec& exArr, int finalSize, int shift, int factor)
{   //default shift = 1, factor = 3
    fvec output(finalSize,0);
    int ind = 0;
    for(int exInd = 0; exInd < exArr.size(); exInd++)
    {
        ind =int((exInd + shift)/factor);

        //choose the best value between all potential values at this index
        output[ind] = std::max(output[ind],exArr[exInd]);
    }

    return output;
}